blockchain/interceptors.go

```diff
@@ -16,29 +28,65 @@ func (p Processor) jobValidationInterceptor(srv interface{}, ss grpc.ServerStrea
 		return status.Errorf(codes.InvalidArgument, "missing metadata")
 	}
 
-	jobAddressMd, ok := md[JobAddressHeader]
-	if !ok {
-		return status.Errorf(codes.InvalidArgument, "missing snet-job-address")
+	paymentHandler, err := p.newPaymentHandlerByType(md)
+	if err != nil {
+		return err
 	}
 
-	jobAddressBytes := common.FromHex(jobAddressMd[0])
+	err = paymentHandler.validatePayment()
+	if err != nil {
+		return err
+	}
 
-	jobSignatureMd, ok := md[JobSignatureHeader]
-	if !ok {
-		return status.Errorf(codes.InvalidArgument, "missing snet-job-signature")
+	err = handler(srv, ss)
+
+	return paymentHandler.completePayment(err)
+}
+
+func (p Processor) newPaymentHandlerByType(md metadata.MD) (paymentHandlerType, error) {
+	paymentTypeMd, ok := md[PaymentTypeHeader]
+
+	paymentType := JobPaymentType
+	if ok && len(paymentTypeMd) > 0 {
+		paymentType = paymentTypeMd[0]
 	}
 
-	jobSignatureBytes := common.FromHex(jobSignatureMd[0])
+	switch {
+	case paymentType == JobPaymentType:
+		return newJobPaymentHandler(&p, md), nil
+	case paymentType == EscrowPaymentType:
+		return newEscrowPaymentHandler(), nil
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unexpected \"%v\", value: \"%v\"", PaymentTypeHeader, paymentType)
+	}
+}
 
-	if p.IsValidJobInvocation(jobAddressBytes, jobSignatureBytes) {
-		err := handler(srv, ss)
-		if err == nil {
-			p.CompleteJob(jobAddressBytes, jobSignatureBytes)
-		}
-		return err
+type paymentHandlerType interface {
+	validatePayment() error
+	completePayment(error) error
```

Comment:
It would be good to have an interface where all functions take an error as an argument of neither of them.